h2h/research/four-factor-model/final_pipeline.ipynb

```diff
@@ -0,0 +1,1538 @@
+       "19394      NaN      NaN      NaN            NaN         NaN          NaN   \n",
+       "19395      NaN      NaN      NaN            NaN         NaN          NaN   \n",
+       "19396      NaN      NaN      NaN            NaN         NaN          NaN   \n",
+       "19397      NaN      NaN      NaN            NaN         NaN          NaN   \n",
+       "19398      NaN      NaN      NaN            NaN         NaN          NaN   \n",
+       "\n",
+       "      result  \n",
+       "0        NaN  \n",
+       "1        NaN  \n",
+       "2        NaN  \n",
+       "3        NaN  \n",
+       "4        NaN  \n",
+       "...      ...  \n",
+       "19394    NaN  \n",
+       "19395    NaN  \n",
+       "19396    NaN  \n",
+       "19397    NaN  \n",
+       "19398    NaN  \n",
+       "\n",
+       "[19399 rows x 13 columns]"
+      ]
+     },
+     "execution_count": 151,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "unencoded_df"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 152,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "21201214 float division by zero\n",
+      "21201214 float division by zero\n"
+     ]
+    }
+   ],
+   "source": [
+    "bad_data = []\n",
+    "for index, (game_id, obj) in enumerate(game_info_dict.items()):\n",
+    "    winner_id = obj[\"winner\"]\n",
+    "    home_dict = obj[\"home\"]\n",
+    "    away_dict = obj[\"away\"]\n",
+    "    unencoded_df.at[index, \"result\"] = 1 if winner_id == home_dict[\"team_id\"] else 0\n",
+    "    \n",
+    "    for dic in [home_dict, away_dict]:\n",
+    "        loc_prefix = \"hm_\" if dic == home_dict else \"aw_\"\n",
+    "        fg = 0\n",
+    "        threep = 0\n",
+    "        fga = 0\n",
+    "        tov = 0\n",
+    "        oreb = 0\n",
+    "        dreb = 0\n",
+    "        ftm = 0   \n",
+    "        for player_name, player_obj in dic[\"players\"].items():\n",
+    "            try: #this is an individual player loop.... if a single player can't be accessed, its ok to skip him\n",
+    "                fg += player_obj[\"stats\"][\"FGM\"]\n",
+    "                threep += player_obj[\"stats\"][\"FG3M\"]\n",
+    "                fga += player_obj[\"stats\"][\"FGA\"]\n",
+    "                tov += player_obj[\"stats\"][\"TOV\"]\n",
+    "                oreb += player_obj[\"stats\"][\"OREB_PCT\"]\n",
+    "                dreb += player_obj[\"stats\"][\"DREB_PCT\"]\n",
+    "                ftm += player_obj[\"stats\"][\"FTM\"]\n",
+    "            except Exception as e:\n",
+    "                print(player_name, player_obj, e)\n",
+    "        try: #this is at a team level... if we fail here, we should not have this trianing example\n",
+    "            unencoded_df.at[index, loc_prefix + \"shooting\"] = (fg+0.5*threep)/fga\n",
```

Comment:
You can make an array of loc_prefix + [stuff] so you don't need to keep generating it every time